cmd/root.go

```diff
@@ -129,6 +131,35 @@ func getEnabledChecks(sp *spol.ScorecardPolicy, argsChecks []string) (checker.Ch
 	return enabledChecks, nil
 }
 
+func createRepoClient(ctx context.Context, uri *repos.RepoURI) (clients.RepoClient, *zap.Logger, error) {
+	var rc clients.RepoClient
+	switch uri.GetType() {
+	// URL.
+	case repos.RepoTypeURL:
+		if err := repo.IsValidGitHubURL(); err != nil {
+			return rc, nil, sce.WithMessage(sce.ErrScorecardInternal, err.Error())
+		}
+		logger, err := githubrepo.NewLogger(*logLevel)
+		if err != nil {
+			return rc, nil, sce.WithMessage(sce.ErrScorecardInternal, err.Error())
+		}
+		// Warning: nothing must fail in this function after logger is constructed.
+		rc = githubrepo.CreateGithubRepoClient(ctx, logger)
+		return rc, logger, nil
+
+	// Local directory.
+	case repos.RepoTypeLocalDir:
+		logger, err := githubrepo.NewLogger(*logLevel)
+		if err != nil {
+			return rc, nil, sce.WithMessage(sce.ErrScorecardInternal, err.Error())
+		}
```

Comment:
Hmm, why is it not natural to call `defer logger.Close()` at the caller site? For example, when handling an HTTP request in Go, the convention is:

```
resp, _ := http.Get("")
defer resp.Body.Close()
if err := json.Unmarshal(resp.Body); err != nil {
  panic(err)
}
```

Which is similar to `logger.Close()` usage. FYI, it's only a nitpicky suggestion. You can leave it as-is if you want (without the warning comment).

Reply:
sorry, I confused you. You're correct about `defer`. The problem is that the `resp` needs to be returned. In our case the function returns an error as well, so the caller would do:
```
resp, err := http.Get("")
if err {
   return blabla
}
defer resp.Body.Close()
```

this would never call the `Close()`. That's why I create `logger` and return it only with an `error=nil`. Is this a better explanation?